src/Catalyst.Core.Modules.Dfs/UnixFs/UnixFs.cs

```diff
@@ -0,0 +1,136 @@
+using Catalyst.Abstractions.Keystore;
+using Catalyst.Core.Lib.Dag;
+using Lib.P2P;
+using ProtoBuf;
+
+namespace Catalyst.Core.Modules.Dfs.UnixFs
+{
+    /// <summary>
+    ///   Support for the *nix file system.
+    /// </summary>
+    internal class UnixFs
+    {
+        private static readonly byte[] EmptyData = new byte[0];
+
+        /// <summary>
+        ///   Creates a stream that can read the supplied <see cref="Cid"/>.
+        /// </summary>
+        /// <param name="id">
+        ///   The identifier of some content.
+        /// </param>
+        /// <param name="blockService">
+        ///   The source of the cid's data.
+        /// </param>
+        /// <param name="keyChain">
+        ///   Used to decypt the protected data blocks.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's value is
+        ///   a <see cref="Stream"/> that produces the content of the <paramref name="id"/>.
+        /// </returns>
+        /// <remarks>
+        ///  The id's <see cref="Cid.ContentType"/> is used to determine how to read
+        ///  the conent.
+        /// </remarks>
+        public static Task<Stream> CreateReadStreamAsync(Cid id,
+            IBlockApi blockService,
+            IKeyApi keyChain,
+            CancellationToken cancel)
+        {
+            switch (id.ContentType)
+            {
+                // TODO: A content-type registry should be used.
+                case "dag-pb":
+                    return CreateDagProtoBufStreamAsync(id, blockService, keyChain, cancel);
+                case "raw":
+                    return CreateRawStreamAsync(id, blockService, cancel);
+                case "cms":
+                    return CreateCmsStreamAsync(id, blockService, keyChain, cancel);
+                default:
+                    throw new NotSupportedException($"Cannot read content type '{id.ContentType}'.");
+            }
+        }
+
+        private static async Task<Stream> CreateRawStreamAsync(Cid id,
+            IBlockApi blockService,
+            CancellationToken cancel)
+        {
+            var block = await blockService.GetAsync(id, cancel).ConfigureAwait(false);
+            return block.DataStream;
+        }
+
+        private static async Task<Stream> CreateDagProtoBufStreamAsync(Cid id,
+            IBlockApi blockService,
+            IKeyApi keyChain,
+            CancellationToken cancel)
+        {
+            var block = await blockService.GetAsync(id, cancel).ConfigureAwait(false);
+            var dag = new DagNode(block.DataStream);
+            var dm = Serializer.Deserialize<DataMessage>(dag.DataStream);
+
+            if (dm.Type != DataType.File)
+            {
+                throw new Exception($"'{id.Encode()}' is not a file.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)